File: apps/python-api/src/yourmedportal_api/composition/app_container.py
from __future__ import annotations

from sqlalchemy.orm import Session

from yourmedportal_api.application.admin.admin_activity_service import AdminActivityService
from yourmedportal_api.application.admin.admin_auth_service import AdminAuthService
from yourmedportal_api.application.admin.admin_booking_service import AdminBookingService
from yourmedportal_api.application.admin.admin_macro_area_service import AdminMacroAreaService
from yourmedportal_api.application.admin.admin_report_service import AdminReportService
from yourmedportal_api.application.admin.admin_report_upload_service import AdminReportUploadService
from yourmedportal_api.application.admin.admin_user_service import AdminUserService
from yourmedportal_api.application.admin.admin_visit_type_service import AdminVisitTypeService
from yourmedportal_api.application.public.public_booking_service import PublicBookingService
from yourmedportal_api.application.public.public_catalog_service import PublicCatalogService
from yourmedportal_api.application.public.public_report_service import PublicReportService
from yourmedportal_api.application.public.public_scheduling_service import PublicSchedulingService
from yourmedportal_api.application.support.activity_logger import ActivityLogger
from yourmedportal_api.application.support.auth_service import AuthService
from yourmedportal_api.infrastructure.repositories.activity_log_repository import ActivityLogRepository
from yourmedportal_api.infrastructure.repositories.admin_user_repository import AdminUserRepository
from yourmedportal_api.infrastructure.repositories.booking_repository import BookingRepository
from yourmedportal_api.infrastructure.repositories.macro_area_repository import MacroAreaRepository
from yourmedportal_api.infrastructure.repositories.report_attachment_repository import ReportAttachmentRepository
from yourmedportal_api.infrastructure.repositories.report_repository import ReportRepository
from yourmedportal_api.infrastructure.repositories.visit_type_repository import VisitTypeRepository


class AppContainer:
    def __init__(self, session: Session, upload_dir: str) -> None:
        macro_areas = MacroAreaRepository(session)
        visit_types = VisitTypeRepository(session)
        bookings = BookingRepository(session)
        reports = ReportRepository(session)
        report_attachments = ReportAttachmentRepository(session)
        admin_users = AdminUserRepository(session)
        activity_logs = ActivityLogRepository(session)
        activity = ActivityLogger(activity_logs)
        auth = AuthService()

        self.auth = auth
        self.public_catalog = PublicCatalogService(macro_areas, visit_types)
        self.public_scheduling = PublicSchedulingService(visit_types, macro_areas, bookings)
        self.public_booking = PublicBookingService(visit_types, macro_areas, bookings)
        self.public_report = PublicReportService(reports, visit_types, report_attachments)
        self.admin_auth = AdminAuthService(admin_users, auth, activity, bookings, reports)
        self.admin_macro_area = AdminMacroAreaService(macro_areas, activity)
        self.admin_visit_type = AdminVisitTypeService(visit_types, activity)
        self.admin_booking = AdminBookingService(bookings, visit_types, macro_areas, activity)
        self.admin_report = AdminReportService(reports, activity)
        self.admin_activity = AdminActivityService(activity_logs)
        self.admin_user = AdminUserService(admin_users, activity)
        self.admin_report_upload = AdminReportUploadService(
            reports, report_attachments, activity, upload_dir
        )

File: apps/api/src/composition/app-container.ts
import { join as pathJoin } from 'node:path';
import { db } from '../db/index.js';
import { ActivityLogRepository } from '../infrastructure/repositories/activity-log-repository.js';
import { AdminUserRepository } from '../infrastructure/repositories/admin-user-repository.js';
import { BookingRepository } from '../infrastructure/repositories/booking-repository.js';
import { MacroAreaRepository } from '../infrastructure/repositories/macro-area-repository.js';
import { ReportAttachmentRepository } from '../infrastructure/repositories/report-attachment-repository.js';
import { ReportRepository } from '../infrastructure/repositories/report-repository.js';
import { VisitTypeRepository } from '../infrastructure/repositories/visit-type-repository.js';
import { AdminActivityService } from '../application/admin/admin-activity-service.js';
import { AdminAuthService } from '../application/admin/admin-auth-service.js';
import { AdminBookingService } from '../application/admin/admin-booking-service.js';
import { AdminMacroAreaService } from '../application/admin/admin-macro-area-service.js';
import { AdminReportService } from '../application/admin/admin-report-service.js';
import { AdminReportUploadService } from '../application/admin/admin-report-upload-service.js';
import { AdminUserService } from '../application/admin/admin-user-service.js';
import { AdminVisitTypeService } from '../application/admin/admin-visit-type-service.js';
import { PublicBookingService } from '../application/public/public-booking-service.js';
import { PublicCatalogService } from '../application/public/public-catalog-service.js';
import { PublicReportService } from '../application/public/public-report-service.js';
import { PublicSchedulingService } from '../application/public/public-scheduling-service.js';
import { ActivityLogger } from '../application/support/activity-logger.js';
import { AuthService } from '../application/support/auth-service.js';

export class AppContainer {
	readonly auth: AuthService;
	readonly publicCatalog: PublicCatalogService;
	readonly publicScheduling: PublicSchedulingService;
	readonly publicBooking: PublicBookingService;
	readonly publicReport: PublicReportService;
	readonly adminAuth: AdminAuthService;
	readonly adminMacroArea: AdminMacroAreaService;
	readonly adminVisitType: AdminVisitTypeService;
	readonly adminBooking: AdminBookingService;
	readonly adminReport: AdminReportService;
	readonly adminActivity: AdminActivityService;
	readonly adminUser: AdminUserService;
	readonly adminReportUpload: AdminReportUploadService;

	constructor(uploadDir: string) {
		const macroAreas = new MacroAreaRepository(db);
		const visitTypes = new VisitTypeRepository(db);
		const bookings = new BookingRepository(db);
		const reports = new ReportRepository(db);
		const reportAttachments = new ReportAttachmentRepository(db);
		const adminUsers = new AdminUserRepository(db);
		const activityLogs = new ActivityLogRepository(db);
		const activity = new ActivityLogger(activityLogs);
		const auth = new AuthService();

		this.auth = auth;
		this.publicCatalog = new PublicCatalogService(macroAreas, visitTypes);
		this.publicScheduling = new PublicSchedulingService(visitTypes, macroAreas, bookings);
		this.publicBooking = new PublicBookingService(visitTypes, macroAreas, bookings);
		this.publicReport = new PublicReportService(reports, visitTypes, reportAttachments);
		this.adminAuth = new AdminAuthService(adminUsers, auth, activity, bookings, reports);
		this.adminMacroArea = new AdminMacroAreaService(macroAreas, activity);
		this.adminVisitType = new AdminVisitTypeService(visitTypes, activity);
		this.adminBooking = new AdminBookingService(bookings, visitTypes, macroAreas, activity);
		this.adminReport = new AdminReportService(reports, activity);
		this.adminActivity = new AdminActivityService(activityLogs);
		this.adminUser = new AdminUserService(adminUsers, activity);
		this.adminReportUpload = new AdminReportUploadService(reports, reportAttachments, activity, uploadDir);
	}
}

export function createDefaultAppContainer() {
	return new AppContainer(pathJoin(process.cwd(), 'uploads'));
}
